src/record_ts.cpp

```diff
@@ -0,0 +1,1446 @@
+        this->data_size = this->alloc_size;
+    }
+
+    /**
+     * Calculate the number of elements given some variable type.
+     */
+    uint32_t size(VarType vtype) { return size(type_size[(uint32_t) vtype]); }
+    /**
+     * Calculate the number of elements given some type size.
+     * This depends on how an operation accesses a variable.
+     * For example, a memcpy operation might access a variable as an array of
+     * \ref uint8_t types, whereas a kernel can access the same variable as an
+     * array of \ref uint64_t.
+     * This changes the size of the variable when inferring the size of the
+     * kernel launch.
+     */
+    uint32_t size(uint32_t tsize) {
+        if (tsize == 0)
+            jitc_fail("replay(): Invalid var type!");
+        size_t size = (this->data_size / (size_t) tsize);
+
+        if (size == 0)
+            jitc_fail("replay(): Error, determining size of variable! init "
+                      "%u, dsize=%zu",
+                      (uint32_t) this->init, this->data_size);
+
+        if (size * (size_t) tsize != this->data_size)
+            jitc_fail("replay(): Error, determining size of variable!");
+
+        return (uint32_t) size;
+    }
+
+    void alloc(JitBackend backend, uint32_t size, VarType type) {
+        alloc(backend, size, type_size[(uint32_t) type]);
+    }
+    void alloc(JitBackend backend, uint32_t size, uint32_t isize) {
+        size_t dsize = ((size_t) size) * ((size_t) isize);
+        return alloc(backend, dsize);
+    }
+    /**
+     * Allocates the data for this replay variable.
+     * If this is attempted twice, we test weather the allocated size is
+     * sufficient and re-allocate the memory if necessary.
+     */
+    void alloc(JitBackend backend, size_t dsize) {
+        AllocType alloc_type =
+            backend == JitBackend::CUDA ? AllocType::Device : AllocType::Host;
+
+        if (!data) {
+            this->alloc_size = dsize;
+            jitc_log(LogLevel::Debug, "    allocating output of size %zu.",
+                     dsize);
+            if (!dry_run)
+                data = jitc_malloc(alloc_type, dsize);
+        } else if (this->alloc_size < dsize) {
+            this->alloc_size = dsize;
+            jitc_log(LogLevel::Debug, "    re-allocating output of size %zu.",
+                     dsize);
+            if (!dry_run) {
+                jitc_free(this->data);
+                data = jitc_malloc(alloc_type, dsize);
+            }
+        } else {
+            // Do not reallocate if the size is enough
+        }
+
+        this->data_size = dsize;
+    }
+
+    void free() {
+        jitc_free(this->data);
+        this->data       = nullptr;
+        this->data_size  = 0;
+        this->alloc_size = 0;
+    }
+};
+
+/// Kernel parameter buffer
+static std::vector<void *> kernel_params;
+
+/// Temporary variables used for replaying a recording.
+static std::vector<ReplayVariable> replay_variables;
```

Comment:
What happens if someone tries replaying two different recordings from two different threads?
(I'm not sure what thread-safety the new DrJit version provides)

Reply:
Its only accessed in `jitc_*` functions, which are behind the `state.lock` mutex.